runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/status/WorkerStatusClient.java

```diff
@@ -0,0 +1,165 @@
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Consumer;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi.WorkerStatusRequest;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi.WorkerStatusResponse;
+import org.apache.beam.sdk.fn.IdGenerator;
+import org.apache.beam.sdk.fn.IdGenerators;
+import org.apache.beam.sdk.fn.stream.SynchronizedStreamObserver;
+import org.apache.beam.vendor.grpc.v1p21p0.io.grpc.stub.StreamObserver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Client for handling requests and responses over Fn Worker Status Api between runner and SDK
+ * Harness.
+ */
+class WorkerStatusClient implements Closeable {
+
+  public static final Logger LOG = LoggerFactory.getLogger(WorkerStatusClient.class);
+  private final IdGenerator idGenerator = IdGenerators.incrementingLongs();
+  private final StreamObserver<WorkerStatusRequest> requestReceiver;
+  private final Map<String, CompletableFuture<WorkerStatusResponse>> responseQueue =
+      new ConcurrentHashMap<>();
+  private final String workerId;
+  private Consumer<String> deregisterCallback;
+  private AtomicBoolean isClosed = new AtomicBoolean(false);
+
+  private WorkerStatusClient(String workerId, StreamObserver<WorkerStatusRequest> requestReceiver) {
+    this.requestReceiver = SynchronizedStreamObserver.wrapping(requestReceiver);
+    this.workerId = workerId;
+  }
+
+  /**
+   * Create new status api client with SDK Harness worker id and request observer.
+   *
+   * @param workerId SDK Harness worker id.
+   * @param requestObserver The outbound request observer this client uses to send new status
+   *     requests to its corresponding SDK Harness.
+   * @return {@link WorkerStatusClient}
+   */
+  public static WorkerStatusClient forRequestObserver(
+      String workerId, StreamObserver<WorkerStatusRequest> requestObserver) {
+    return new WorkerStatusClient(workerId, requestObserver);
+  }
+
+  /**
+   * Get the latest sdk worker status from the client's corresponding SDK Harness. A random id will
+   * be used to specify the request_id field.
+   *
+   * @return {@link CompletableFuture} of the SDK Harness status response.
+   */
+  public CompletableFuture<WorkerStatusResponse> getWorkerStatus() {
+    WorkerStatusRequest request =
+        WorkerStatusRequest.newBuilder().setId(idGenerator.getId()).build();
+    return getWorkerStatus(request);
+  }
+
+  /**
+   * Get the latest sdk worker status from the client's corresponding SDK Harness with request.
+   *
+   * @param request WorkerStatusRequest to be sent to SDK Harness.
+   * @return {@link CompletableFuture} of the SDK Harness status response.
+   */
+  CompletableFuture<WorkerStatusResponse> getWorkerStatus(WorkerStatusRequest request) {
+    CompletableFuture<WorkerStatusResponse> future = new CompletableFuture<>();
```

Comment:
You have a concurrency race between `getWorkerStatus` adding to the `responseQueue` and `close()`

Reply:
Added synchronization.

Reply:
This race still exists since adding to `pendingResponses` isn't synchronized with the iteration in `close()`

You want to wrap pending responses mutation with synchronization:
```
synchronized (pendingResponses) {
  this.pendingResponses.put(...);
}
```

Reply:
wouldn't synchronizedMap implicitly call the synchronize function?

Reply:
Your right because the javadoc does say that Collections.synchronousMap synchronizes on the map delegate so this is not needed.